lib/src/native/realm_core.dart

```diff
@@ -2907,45 +2907,63 @@ extension on Pointer<realm_value_t> {
+  }
+}
 
-    switch (ref.type) {
+extension on realm_value_t {
+  Object? toDartValueByRef(Realm? realm) {
+    switch (type) {
       case realm_value_type.RLM_TYPE_NULL:
         return null;
       case realm_value_type.RLM_TYPE_INT:
-        return ref.values.integer;
+        return values.integer;
       case realm_value_type.RLM_TYPE_BOOL:
-        return ref.values.boolean;
+        return values.boolean;
       case realm_value_type.RLM_TYPE_STRING:
-        return ref.values.string.data.cast<Utf8>().toRealmDartString(length: ref.values.string.size)!;
+        return values.string.data.cast<Utf8>().toRealmDartString(length: values.string.size)!;
       case realm_value_type.RLM_TYPE_FLOAT:
-        return ref.values.fnum;
+        return values.fnum;
       case realm_value_type.RLM_TYPE_DOUBLE:
-        return ref.values.dnum;
+        return values.dnum;
       case realm_value_type.RLM_TYPE_LINK:
-        final objectKey = ref.values.link.target;
-        final classKey = ref.values.link.target_table;
+        if (realm == null) {
+          return null;
+        }
+        final objectKey = values.link.target;
+        final classKey = values.link.target_table;
         if (realm.metadata.getByClassKeyIfExists(classKey) == null) return null; // temprorary workaround to avoid crash on assertion
         return realmCore._getObject(realm, classKey, objectKey);
       case realm_value_type.RLM_TYPE_BINARY:
         throw Exception("Not implemented");
       case realm_value_type.RLM_TYPE_TIMESTAMP:
-        final seconds = ref.values.timestamp.seconds;
-        final nanoseconds = ref.values.timestamp.nanoseconds;
+        final seconds = values.timestamp.seconds;
+        final nanoseconds = values.timestamp.nanoseconds;
         return DateTime.fromMicrosecondsSinceEpoch(seconds * _microsecondsPerSecond + nanoseconds ~/ _nanosecondsPerMicrosecond, isUtc: true);
       case realm_value_type.RLM_TYPE_DECIMAL128:
-        var decimal = ref.values.decimal128; // NOTE: Does not copy the struct!
+        var decimal = values.decimal128; // NOTE: Does not copy the struct!
         decimal = _realmLib.realm_dart_decimal128_copy(decimal); // This is a workaround to that
         return Decimal128Internal.fromNative(decimal);
       case realm_value_type.RLM_TYPE_OBJECT_ID:
-        return ObjectId.fromBytes(cast<Uint8>().asTypedList(12));
+        return ObjectId.fromBytes(values.object_id.bytes.toIntList(ObjectId.byteLength));
       case realm_value_type.RLM_TYPE_UUID:
-        return Uuid.fromBytes(cast<Uint8>().asTypedList(16).buffer);
+        return Uuid.fromBytes(values.uuid.bytes.toIntList(16).buffer);
       default:
-        throw RealmException("realm_value_type ${ref.type} not supported");
+        throw RealmException("realm_value_type $type not supported");
     }
   }
 }
 
+extension on Array<Uint8> {
+  Uint8List toIntList(int count) {
+    List<int> result = List.filled(count, this[0]);
```

Comment:
Why pre-fill the result array?

Reply:
I don't think it is. Also we should not be doing micro optimizations unless we are sure we have a performance issue here.

Reply:
I returned back the pre-filled array, otherwise an error occurs about ´Instance of '_GrowableList'

Reply:
No, it doesn't work. You can see the old code from this commit https://github.com/realm/realm-dart/pull/1291/commits/c37d06175ab3372197c3c4be057e51f3f2e211b8. It is the same like suggested by you, but it throws ´Instance of '_GrowableList' exception.

Reply:
ok,  I think we are now discussing that adding elements to a list of integers is not working in Dart.

Can you try the sample for adding elements to a list in Dart that I posted?

Reply:
Ah, I'm sorry! The issue was with the wrong index `var i = 1`. The length was wrong and the exception was `ArgumentError (Invalid argument (bytes): Bytes array should has length equal to 12: Instance(length:11) of '_GrowableList')`.
I've fixed this.